### PR TITLE
87819: Handle throttling exceptions

### PR DESCRIPTION
**Pull request by zoiec**
Adjust the retry logic to handle the exceptions so we can better see how this is failing (potentially if the failures continue).

ADO: https://dev.azure.com/AMA-Ent/AMA-Ent/_workitems/edit/87819

### CODE REVIEW

**Review by AnnelieseA — DISMISSED** [review]

File: push/cloudwatch/event.rb [hunk 1]
```diff
@@ -23,11 +23,20 @@ def retry_opts
...
+        retries ||= 0
+
         Aws::Lambda::Client.new(retry_opts.merge(opts)).list_tags(
           resource: 'arn:aws:lambda:' + ENV.fetch('REGION') + ':' + owner + ':function:' + log_group
         ).tags
+      rescue Aws::Lambda::Errors::ThrottlingException => e
+        puts "ThrottlingExceptionGetLambdaTags (retry attempt: #{retries}): #{e.inspect}"
+        retry if (retries += 1) < 3
+
+        raise StandardError, 'ThrottlingExceptionRetryLimitExceeded'
```
Comment (zoiec) [inline push/cloudwatch/event.rb]: I don't think we should be raising any error, as that will blow it up

File: subscribe/lambda/function.rb [hunk 2]
```diff
@@ -107,12 +115,19 @@ def cloudwatch_log_group_name
...
             log_group_name: cloudwatch_log_group_name,
             filter_name_prefix: FILTER_NAME,
             limit: 1
           ).subscription_filters
         end
+      rescue Aws::CloudWatchLogs::Errors::ThrottlingException => e
+        puts "ThrottlingExceptionSubscriptions (retry attempt: #{retries}): #{e.inspect}"
+        retry if (retries += 1) < 3
+
+        raise StandardError, 'ThrottlingExceptionRetryLimitExceeded'
```
Comment (zoiec) [inline subscribe/lambda/function.rb]: I don't think we should be raising any error, as that will blow it up

**Review by zoiec — COMMENTED** [review]
A couple of comments

File: spec/push/cloudwatch/event_spec.rb [hunk 3]
```diff
@@ -36,6 +36,19 @@
...
       end
+
+      context 'when throttling exception is thrown' do
+        before(:each) do
+          allow(Aws::Lambda::Client).to receive(:new).and_raise(
+            Aws::Lambda::Errors::ThrottlingException.new('test', 'test')
+          )
+        end
+
+        it 'raises exception when retry limit exceeded' do
```
Comment (zoiec) [inline spec/push/cloudwatch/event_spec.rb]: Could you update the description

File: spec/subscribe/lambda/function_spec.rb [hunk 4]
```diff
@@ -254,23 +252,18 @@
...
+        before(:each) do
+          allow(cloudwatch).to receive(:describe_subscription_filters).and_raise(
+            Aws::CloudWatchLogs::Errors::ThrottlingException.new('test', 'test')
+          )
+        end
 
-      it 'returns a hash after multiple throttle errors' do
-        response = described_class.subscribe_all!(lambda, cloudwatch)
-        expect(response).to be_a(Hash)
+        it 'raises exception when retry limit exceeded' do
```
Comment (zoiec) [inline spec/subscribe/lambda/function_spec.rb]: Could you update the description

File: spec/subscribe/lambda/function_spec.rb [hunk 5]
```diff
@@ -44,24 +44,22 @@
...
-            tags_stub
-          ]
+        allow(lambda).to receive(:list_tags).and_raise(
+          Aws::Lambda::Errors::ThrottlingException.new('test', 'test')
         )
       end
 
-      it 'returns a Subscribe::Lambda::Function after multiple throttle errors' do
-        expect(subject).to be_a(Subscribe::Lambda::Function)
+      it 'raises exception when retry limit exceeded' do
```
Comment (zoiec) [inline spec/subscribe/lambda/function_spec.rb]: Could you update the description

**Review by zoiec — COMMENTED** [review]
image

**Review by dowied4 — APPROVED** [review]

**Pull request merged**